docs/proposals/auth-gateway.md

```diff
@@ -0,0 +1,67 @@
+---
+title: "Authentication Gateway"
+linkTitle: "Authentication Gateway"
+weight: 1
+slug: "auth-gateway"
+---
+
+- Author: [Doğukan Teber](https://github.com/dogukanteber)
+- Date: March 2023
+- Status: Proposed
+
+## Overview
+
+If you run Cortex for multiple tenants you need to identify your tenants every time they send metrics or query them. This is needed to ensure that metrics can be ingested and queried separately from each other. For this purpose, the Cortex microservices require you to pass a Header called `X-Scope-OrgID`. Unfortunately, the Prometheus Remote write API has no config option to send headers and for Grafana you must provide a data source to do so. Therefore the Cortex k8s manifests suggest deploying an NGINX server inside of each tenant which acts as a reverse proxy. Its sole purpose is proxying the traffic and setting the `X-Scope-OrgID` header for your tenant.
+
+## Proposal
+
+We propose to solve this problem by adding a gateway that can be considered the entry point for all requests towards Cortex. This gateway will be responsible for handling multi-tenancy features without complicated proxy configuration.
+
+This gateway will be located in another repository, [cortexproject/auth-gateway](https://github.com/cortexproject/auth-gateway). For the first version, basic authentication will be supported. However, as the project grows, it is planned to support different authentication mechanisms according to the needs of the users.
+
+We would like to have a simple configuration format. The following is the proposed configuration:
+
+```yaml
+server:
+  address: localhost
+  port: 8080
+admin:
+  address: localhost
+  port: 8090
+tenants:
+  - authentication: basic
+    username: username1
+    password: password1
+    id: orgid
+
+  - authentication: basic
+    username: username2
+    password: password2
+    id: orgid
+distributor:
+  dns_refresh_interval: 3s
+  url: http://distributor:9009
+  paths:
+  - /api/v1/push
+  - /api/prom/push
+  http_client_timeout: 20
+  read_timeout: 5
+  write_timeout: 5
+  idle_timeout: 7
+frontend:
+  url: http://frontend:9010
+  read_timeout: 5
+  write_timeout: 5
+  idle_timeout: 7
+
+...
+
+```
+
+`admin` will be used for endpoints that do not require authentication, ie. /metrics, /ready, /pprof. If no paths are given in the component, such as `frontend`, all of the endpoints of that component will be registered. Additionally, the users will be able to specify custom timeouts for each component similar to [NGINX](https://github.com/cortexproject/cortex-helm-chart/blob/571fc2a5f184b6b7c243bac3727503264249bfd1/templates/nginx/nginx-config.yaml#L50-L55).
```

Comment:
> `admin` will be used for endpoints that do not require authentication, ie. /metrics, /ready, /pprof

Is this a good idea? Do we need `admin` at all from the gateway perspective? Does it make sense to load balance for `/metrics`, `/ready`, and `/pprof` endpoints at all? One typically need to know which pods they are talking to for these endpoints right?

Reply:
I suggested to create the admin port. Typically in prometheus, the [/metrics](https://cortexmetrics.io/docs/api/#metrics) endpoint has no auth, it's considered a trusted endpoint. The [/ready ](https://cortexmetrics.io/docs/api/#readiness-probe) endpoint is used by kubernetes to verify if a pod is ready. same for [pprof](https://cortexmetrics.io/docs/api/#pprof). (Which is /debug/pprof btw)
These endpoints typically have no auth and are already provided in all the cortex components without auth (There is no change)

There is no load balance required for these endpoints, it's just for admin purposes on the single containers.

>One typically need to know which pods they are talking to for these endpoints right?
Not typically, no.

Having a separate port for non-authenticated requests makes it simple for admins to know what port to protect.

Reply:
I see make sense. 

One last question regarding: 

> There is no load balance required for these endpoints, it's just for admin purposes on the single containers.

Does that mean the auth-gateway will need to support persistent connection? Like if I want to do run multiple `/pprof` on a single container to compare CPU flame graph at two different time instant, how would I do that through auth-gateway?

Reply:
No, I don't think you would use auth-gateway for that. You would connect directly to each pod. Pprof and metrics are not directly made for tenants, but for cortex operators to understand cortex internals, there is no auth or special load balancing needed there IMHO.